### DISCUSSION TITLE
Metal shaders explanation

### DISCUSSION THREAD

**Discussion by Labrium** [zed-industries/zed#14592] [Feature Requests]
As someone affected by the rendering bug described in #5062 and #6853, I decided to investigate a little deeper. I suspect that since the problem seems to be GPU vendor specific, it might be a problem with the drawing pipeline commands or shader behaving incorrectly, probably due to the slightly older Metal drivers for NVIDIA GPUs on macOS.

A bit of code reference following led me to the `monochrome_sprite` shader code, https://github.com/zed-industries/zed/blob/cb6fc11abc343ff146efb9ddb1db8a55e77632e7/crates/gpui/src/platform/mac/shaders.metal#L284-L331

which I noticed is very similar to the `polychrome_sprite` code, https://github.com/zed-industries/zed/blob/cb6fc11abc343ff146efb9ddb1db8a55e77632e7/crates/gpui/src/platform/mac/shaders.metal#L333-L390

which I assume is used to draw emojis (tell me if this is incorrect), and according to https://github.com/zed-industries/zed/issues/5062#issuecomment-1959776995, the rendering issue doesn't happen to emojis.

After comparing the code for both of these, I noticed that they seem to reference `sprite_id` in a similar way in the vertex functions, but the `MonochromeSprite` input and output structs don't include `uint sprite_id [[flat]];` like the `PolychromeSprite` structs do.

I don't have any experience with Metal, but I have worked with OpenGL ES for a while, and I have an idea of how the modern graphics pipeline and unified shader model works. Can someone explain how `sprite_id` is used, why it is declared in one set of structs but not the other, and if this could be related to the NVIDIA-specific rendering bug?

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
